### PR TITLE
list of goods(hooks)

### PR DESCRIPTION
**Pull request by Ambulence**
-[DEMO LINK](https://Ambulence.github.io/react_list-of-goods-hooks/)

### CODE REVIEW

**Comment by nazarovoleksiy** [comment]
Good Job

**Review by nazarovoleksiy — APPROVED** [review]
Good Job

File: src/App.tsx [hunk 1]
```diff
@@ -1,63 +1,117 @@
...
+        return firstGoods.length - secondGoods.length;
+
+      default:
+        return 0;
+    }
+  });
+
+  if (isReversed) {
+    visibleGoods.reverse();
+  }
```
Comment (Andr1yK) [inline src/App.tsx:37]: u you can reduce the complexity and make reverse inside sort function)

File: src/App.tsx [hunk 2]
```diff
@@ -1,63 +1,117 @@
...
+  const sortByAlphabet = () => {
+    setSortType(SortType.ALPHABET);
+  };
+
+  const sortByLength = () => {
+    setSortType(SortType.LENGTH);
+  };
+
+  const reverse = () => {
+    setReverse(!isReversed);
```
Comment (Andr1yK) [inline src/App.tsx:56]: use functional setState

File: src/App.tsx [hunk 3]
```diff
@@ -1,63 +1,117 @@
...
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
-      </ul>
+      <GoodsList goods={visibleGoods} />
+
```
Comment (Andr1yK) [inline src/App.tsx:114]:
```suggestion
      <GoodsList goods={visibleGoods} />
```

File: src/components/GoodsList.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+
+type Props = {
+  goods: string[];
+};
+
+export const GoodsList: React.FC<Props> = ({ goods }) => {
+  return (
```
Comment (Andr1yK) [inline src/components/GoodsList.tsx:8]: useless return

**Review by Andr1yK — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 5]
```diff
@@ -1,63 +1,116 @@
...
+  NONE,
+  ALPHABET,
+  LENGTH,
+}
+
+export function getReorderedGoods(
+  goods: string[],
+  isReversed: boolean,
+  sortType: SortType,
+) {
```
Comment (Andr1yK) [inline src/App.tsx:19]: When working with functions that have a lot of arguments, it can sometimes be difficult to remember the order of the arguments or even to know what each argument is for. To improve the readability and maintainability of your code, it's recommended to group related arguments together in an options object instead.

In the example provided, `isReversed` and `sortType` are two arguments that are related to the sorting of goods. Instead of passing them as separate arguments, we can group them together in an options object:

```js
export interface GoodsSortingOptions {
  isReversed: boolean;
  sortType: SortType;
}

export function getReorderedGoods(
  goods: string[],
  options: GoodsSortingOptions
) {
...
}
```

This approach has several advantages:

- **Readability**: It's easier to understand what each argument is for when they are grouped together in an options object.

- **Flexibility**: We can add new options without breaking existing code. For example, if we want to add a `maxItems` option, we can simply add it to the `GoodsSortingOptions` interface.

- **Reduced cognitive load**: When calling the function, we only need to pass in a single argument instead of multiple arguments.

Here's an example of how we can call the function with the new options object:

```js
const goods = ["apple", "banana", "orange"];
const options = {
  isReversed: true,
  sortType: SortType.ASCENDING,
};

const reorderedGoods = getReorderedGoods(goods, options);
```

In summary, grouping related arguments together in an options object can improve the readability and maintainability of your code, and reduce the cognitive load of calling the function.

File: src/App.tsx [hunk 6]
```diff
@@ -1,63 +1,116 @@
...
+
+  if (isReversed) {
+    visibleGoods.reverse();
+  }
+
+  return visibleGoods;
+}
 
 export const App: React.FC = () => {
+  const [isReversed, setReverse] = useState(false);
```
Comment (Andr1yK) [inline src/App.tsx:43]: In React, when using the `useState` hook to manage state, it's a common convention to name the state variable and its corresponding setter function using a specific naming template: `[variableName, "set" + variableName]`.

For example, in the following code:

```js
const [isReversed, setReverse] = useState(false);
```

`isReversed` is the state variable, and setReverse is the setter function for that variable.

Using this naming convention has several advantages:

- **Readability**: It makes it clear that the setReverse function is specifically for setting the `isReversed` state variable, which can improve the readability of the code.

- **Consistency**: Using a consistent naming convention makes it easier for other developers to understand the codebase and reduces the cognitive load of working with it.

- **Autocompletion**: Many code editors and IDEs will suggest the correct setter function name based on the state variable name, which can make it faster to write code.

Here's an example of how we can use this naming convention with multiple state variables:

```js
const [isReversed, setIsReversed] = useState(false);
const [selectedItem, setSelectedItem] = useState(null);
const [count, setCount] = useState(0);
```

In summary, naming your React state variables and setter functions using the `[variableName, "set" + variableName]` template can improve the readability, consistency, and autocompletion of your code.

File: src/App.tsx [hunk 7]
```diff
@@ -1,63 +1,116 @@
...
         </button>
 
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
+        {
+          (isReversed || sortType !== SortType.NONE) && (
```
Comment (Andr1yK) [inline src/App.tsx:101]: let's move this condition to const)

**Review by Andr1yK — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 8]
```diff
@@ -1,63 +1,183 @@
...
+//   });
+
+//   return visibleGoods;
+// }
+
+// test code
+export function getReorderedGoods(
+  goods: string[],
+  { isReversed = false, sortType }: GoodsSortingOptions,
+  isReverseSorting = false,
```
Comment (sergii-nosachenko) [inline src/App.tsx:47]: Why do you need this argument?

File: src/App.tsx [hunk 9]
```diff
@@ -1,63 +1,183 @@
...
+        break;
+
+      default:
+        order = 0;
+        break;
+    }
+
+    if (isReverseSorting) {
+      order *= -1;
+    }
```
Comment (sergii-nosachenko) [inline src/App.tsx:70]: I'm not sure this is a good idea.

File: src/App.tsx [hunk 10]
```diff
@@ -1,63 +1,183 @@
...
+//       case SortType.LENGTH:
+//         return (isReversed ? -1 : 1) * (firstGoods.length - secondGoods.length);
+
+//       default:
+//         return 0;
+//     }
+//   });
+
+//   return visibleGoods;
+// }
```
Comment (sergii-nosachenko) [inline src/App.tsx:41]: Clean or use it. You need to decide.

File: src/App.tsx [hunk 11]
```diff
@@ -1,63 +1,183 @@
...
+//         return 0;
+//     }
+//   });
+
+//   if (isReversed) {
+//     visibleGoods.reverse();
+//   }
+
+//   return visibleGoods;
+// }
```
Comment (sergii-nosachenko) [inline src/App.tsx:103]: Same here

File: src/App.tsx [hunk 12]
```diff
@@ -1,63 +1,183 @@
...
+  const reverse = () => {
+    setIsReversed(currentReverse => !currentReverse);
+  };
+
+  const reset = () => {
+    setSortType(SortType.NONE);
+    setIsReversed(false);
+  };
+
+  const showReset = isReversed || sortType !== SortType.NONE;
```
Comment (sergii-nosachenko) [inline src/App.tsx:129]:
```suggestion
  const shouldShowReset = isReversed || sortType !== SortType.NONE;
```

File: src/App.tsx [hunk 13]
```diff
@@ -1,63 +1,183 @@
...
         </button>
 
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
+        {
+          (showReset) && (
```
Comment (sergii-nosachenko) [inline src/App.tsx:168]:
```suggestion
        {showReset && (
```

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]
You have not passed all tests. Check your code.

File: src/App.tsx [hunk 14]
```diff
@@ -1,63 +1,124 @@
...
-  'Honey',
-  'Jam',
-  'Garlic',
-];
+import goodsFromServer from './data/goods.json';
+
+enum SortType {
+  NONE,
+  ALPHABET,
+  LENGTH,
```
Comment (sergii-nosachenko) [inline src/App.tsx:12]: Always add values to enum

**Review by sergii-nosachenko — APPROVED** [review]
Good work
